### ISSUE TITLE
Add a facility for synchronizing concurrent actions

### ISSUE THREAD

**Issue by idanarye** [weka-io/easypy#58]
When running tasks concurrently with `MultiObject`, we often have problems:

* The tasks need to do a certain part in parallel, but reach there at different times.
    * Example: when killing nodes, we want to kill them at the same time (or as closely as possible)
    * Another example: when waiting for nodes to get lease, we want to start waiting together, after all of them were resurrected.
* The tasks need to do something that is expensive or has side-effects or otherwise does not play work that well with concurrency, but the operation and it's result could have been shared between the tasks.
    * Example: writing to the logs - it could be nice to have a single log line(with some aggregated data) instead of separate log line for each concurrent task.
    * Another example: holding an exclusive lock. Sometimes the tasks can share that lock, and just don't want for other things to get it.

**Comment by idanarye** [comment]
# Idea 1 - a synchronization object

```python
def foo(a, b, c, _sync=Sync):
    # bla bla bla

    # Wait for all tasks registered on this sync object to reach this point:
    _sync.wait_for_everyone()

    # 1. Wait for all the tasks to reach this point.
    # 2. Collect the `a` values passed to this into a list(`a_values`)
    # 3. Call _logger.info from a single task (e.g. - the first task to reach this)
    _sync.together(a, lambda a_values: _logger.info('Doing %s %s with %s', b, c, a_values))

    # Alternative syntax - requires some serious black magic (https://stackoverflow.com/a/12594789/794380):
    with _sync.together(a) as a_values:
        _logger.info('Doing %s %s with %s', b, c, a_values)
```

Specifics:
* The `Sync` singleton will provide provide methods that don't actually sync - so the function can be used without concurrency.
* When `MultiObject` / `concurrent_map` (and maybe `concurrent_find` too? not sure if it would make sense there...) need to run a function with a `Sync` argument - they will create a synchronization object, register all the tasks to it, and pass it.
    * Force it to be in an argument named `_sync`, or allow it to be in any argument?
    * Or maybe we should use a decorator on the function as our marker?
* When waiting, the sync method will wait for **all** registered tasks to call the same method from the same line.
* If one of the registered tasks threw, the other tasks will no longer wait for it when they call the sync methods.

Pros:
* Will look the similar to current code, with only the synchronized parts changed.

Cons:
* We can no longer run tasks in batches(with limited number of workers) because then we'll wait for threads that haven't started yet.
    * Unless we can make a worker work on another task while it's first task is waiting?
    * Or maybe gevent can solve this?

**Comment by idanarye** [comment]
# Idea 2 - use the coroutines-async style

We need to support Python 3.4, so we cannot use `async`/`await`, but we can use the old coroutine style for async operations:

```python
@synchronized
def foo(a, b, c):
    # run this part asynchronically

    # Wait for all tasks registered on this sync object to reach this point:
    yield synchronized.wait_for_everyone()

    # 1. Wait for all the tasks to reach this point.
    # 2. Collect the `a` values passed to this into a list(`a_values`)
    # 3. Call _logger.info from a single task (e.g. - the first task to reach this)
    yield synchronized.together(a, lambda a_values: _logger.info('Doing %s %s with %s', b, c, a_values))

    # Alternative syntax - requires some serious black magic (https://stackoverflow.com/a/12594789/794380):
    with (yield synchronized.together(a)) as a_values:
        _logger.info('Doing %s %s with %s', b, c, a_values)
        # Need to make sure to only resume to other tasks after the `__exit__`
```

Specifics:
* Again - when called directly it will behave like a normal function, but `MultiObject` / `concurrent_map` will run it differently. They will run it as generator.
* The "reactor" will call `next`/`send` on all the iterators to receive the _synchronization command_ (`wait_for_everyone` or `together`), make sure it got the same command from all iterators, and perform it.

Pros:
* The non-synchronized parts will run with workers, so we don't have to run everything concurrently to parallelize the parts we want:
    ```python
    # super-simplified node.faults.down()
    @synchronized
    def down(self):
        # pre kill stuff

        yield synchronized.wait_for_everyone()
        self.kill()
        yield synchronized.wait_for_everyone()

        # while-down stuff

        yield synchronized.wait_for_everyone()
        self.revive()
        yield synchronized.wait_for_everyone()

        # post revival stuff
    ```

    If we have 10 works and we want to kill 20 nodes, it'll:
    1. Run the _pre kill stuff_ for the first 10 nodes.
    2. Run the _pre kill stuff_ for the remaining 10 nodes.
    3. Wait for everyone to reach this point.
    4. Kill the first 10 nodes.
    5. Kill the remaining 10 nodes.
    6. Wait for everyone to reach this point.
    7. Do the _while-down stuff_ for the first 10 nodes.
    8. Do the _while-down stuff_ for the remaining 10 nodes.
    9. Wait for everyone to reach this point.
    10. Revive the first 10 nodes.
    11. Revive the remaining 10 nodes.
    12. Wait for everyone to reach this point.
    13. Run the _post revive stuff_ for the first 10 nodes.
    14. Run the _post revive stuff_ for the remaining 10 nodes.

    So we may not have perfect parallelization for the important parts(kill&revive), but at least we do them together (and not, for example, kill one node after we revive another)

Cons:
* What if we want a `@synchronized` generator or context manager? The solution won't be pretty...
